Fix: Validate state when saved display unavailable

MathieuDebit opened the pull request:
## What is this PR

This PR fixes the window positioning at startup when the saved display is unavailable.

See the corresponding issue: https://github.com/mawie81/electron-window-state/issues/44.
Related: https://github.com/mawie81/electron-window-state/issues/1, https://github.com/mawie81/electron-window-state/issues/4, https://github.com/mawie81/electron-window-state/issues/27, https://github.com/mawie81/electron-window-state/issues/31, https://github.com/mawie81/electron-window-state/issues/38, https://github.com/mawie81/electron-window-state/issues/43.

Commit aac92c0 shows that the current implementation is broken since the last test is failing (`Ensure window is visible at startup if saved display is unavailable and was on the left`).

## How does it work
The solution currently adopted consists of reseting the state if the saved display is unavailable.

The advantage is that we ensure the window will always be visible if the saved display is not found. The downside is that the window will not be positioned to the saved coordinates if another display contains the corresponding bounds.

Better implementations that check for each display available and ensure the window is fully visible on one can be found in `electron-boilerplate` and `VS Code`.
See:
- [electron-boilerplate:src/helpers/window.js](https://github.com/szwacz/electron-boilerplate/blob/bd4812720b0bd5f21beb413e567723bae653d5f5/src/helpers/window.js#L58-L68)
- [vscode:src/vs/code/electron-main/window.ts](https://github.com/Microsoft/vscode/blob/95ca89baf72f1e6b3db528533d8bd731e0430103/src/vs/code/electron-main/window.ts#L649-L705)

What about using their methods?

## Test instructions
- Fetch this branch and run `npm link`
- Clone https://github.com/MathieuDebit/electron-quick-start/pull/3
- Run `npm link electron-window-state`
- Follow the reproductions steps on the [pull request](https://github.com/MathieuDebit/electron-quick-start/pull/3) and verify that the bug is fixed

## Todos
- [ ] Add Changelog
- [ ] Bump version

MathieuDebit committed:
Update .gitignore and package-lock.json

MathieuDebit committed:
Add tests to control window visibility at startup

- Ensure window is visible at startup if saved display is unavailable and was on the left or right

MathieuDebit committed:
Fix validateState method when saved display is unavailable

- Reset state to default values on the retrieved display
- Add tests

mawie81 commented:
Thanks a lot for the effort you are putting into this 👍 
I'm fine with simply falling back to the default state in such cases.

I'll do some tests on Windows and Mac tomorrow.

If all is fine and there is nothing else you would like to add in here, I'm happy to merge.
I can the do the Changelog and the version bumping when doing the release 😃

mawie81 commented:
So I did some tests on Mac and Windows. Mac is fine, but on Windows the app is still of screen.
I just positioned the app on the second screen and unplugged the primary one. 
The state before starting again is looking like this:
```js
{ 
  width: 1000,
  height: 326,
  x: 2577,
  y: 673,
  isMaximized: false,
  isFullScreen: false,
  displayBounds: { x: 0, y: 0, width: 1680, height: 1050 } 
}
```
When starting with the single monitor again the bound are fine and the window gets positioned at ```x=2577``` and is off screen.
The strange thing is that the ```displayBounds``` in the state file is retrieved using ```screen.getDisplayMatching(winBounds).bounds``` and the ```winBounds``` in there had ```x=2577``` 🤔 

So the only way a can currently think of fixing this, is the same approach like e.g. electron-boilerplate does. I quickly hacked their ```ensureVisibleOnSomeDisplay``` into the code and it worked.

MathieuDebit commented:
Thanks for your tests 🙏

Actually the bug as described in [my tests](https://github.com/MathieuDebit/electron-quick-start/pull/3) is not reproductible in MacOS because this OS seems to handle out-of-bounds windows differently. See https://github.com/mawie81/electron-window-state/issues/44#issuecomment-438381017. OSX has other issues (like unplug the second screen when the app was fullscreen).

Does your tests pass on Windows? Would you think of a new test that would fail on your case? Unfortunately I can't manage to reproduce your issue 😕 I tried different ways but it's always working for me.

And I added this big general gitignore as it covers pretty all useless file which is nice for open-source projects, but okay I'm removing it!

MathieuDebit committed:
Revert .gitignore

mawie81 commented:
Here is a test that fails and would render the app off screen:
```js
test('test', t => {
  const jsonfile = require('jsonfile');
  sinon.stub(jsonfile, 'readFileSync').returns({
    width: 1000,
    height: 326,
    x: 2577,
    y: 673,
    isMaximized: false,
    isFullScreen: false,
    displayBounds: {x: 0, y: 0, width: 1680, height: 1050}
  });

  const {screen} = require('electron');
  const screenBounds = {x: 0, y: 0, width: 1680, height: 1050};
  sinon.stub(screen, 'getDisplayMatching').returns({bounds: screenBounds});

  const state = require('.')({
    defaultWidth: 500,
    defaultHeight: 300
  });

  t.is(state.x, 0);
  t.is(state.y, 0);
  t.is(state.width, 500);
  t.is(state.height, 300);
  t.is(state.displayBounds, screenBounds);

  jsonfile.readFileSync.restore();
  screen.getDisplayMatching.restore();
});
```

tobiasmuecksch commented:
I've implemented a fix that works fine on windows, but it doesn't work on macOS. Maybe we can merge our solutions somehow. You can find it here: https://github.com/burningparrot/electron-window-state/commit/d906c7f5e68823526ab2e29df212ef29642c796f

MathieuDebit committed:
Add a new failing test

- Validate state if saved display is available but window outside display bounds

MathieuDebit committed:
Add and export resetStateToDefault() method

MathieuDebit committed:
Ensure window is visible on some display during state validation

MathieuDebit commented:
I just pushed a `ensureWindowVisibleOnSomeDisplay()` method that fixes @mawie81's previous issue. Unfortunately there is still a failing test in one particular case 😅

See `Validate state if saved display is available and secondary on left`:
https://github.com/mawie81/electron-window-state/blob/477f3dba797e3ac4139d4263357ae78b40a30e9a/test.js#L256-L292

Still working on it 👨🏻‍🔧

MathieuDebit committed:
Remove deep-equal dependency

tobiasmuecksch commented:
Well, my solution doesn't crash. It's just not working as expected. It always places the window on the primary screen on macOS. I haven't used the built-in tests in this repo yet.

tobiasmuecksch commented on index.js:
```diff
@@ -57,7 +78,9 @@ module.exports = function (options) {
       // Check if the display where the window was last open is still available
       const displayBounds = screen.getDisplayMatching(state).bounds;
       const sameBounds = deepEqual(state.displayBounds, displayBounds, {strict: true});
```
Why do you try to make sure, that the bounds are equal? Isn't that irrelevant?

MathieuDebit committed:
Remove failing test

MathieuDebit commented:
@mawie81 I just removed the last failing test and propose to open another PR to fix it. This way we can merge this one that fixes a lot of issues for now.

Ready for review  ✨

tobiasmuecksch commented:
I agree. What do you think @mawie81 ?

tobiasmuecksch commented:
I'd love to test this, but I don't find any button or something alike to check it out.

MathieuDebit commented:
@tobiasmuecksch The fork is https://github.com/getstation/electron-window-state.
You can clone the branch this way:

```sh
git clone https://github.com/getstation/electron-window-state.git -b fix/validate-state-when-saved-display-unavailable
cd electron-window-state
npm install
npm link
```

Otherwise on can put `"electron-window-state": "github:mawie81/electron-window-state#pull/45/head"` in your `package.json` to install this PR directly ✨

tobiasmuecksch commented:
@MathieuDebit Thank you, I was looking for the URL. Will test it over the weekend.

MathieuDebit commented:
Yes there is this kind of issue on MacOS too! You right we could handle this in another issue, let's validate this PR first 🎉

tobiasmuecksch commented:
@MathieuDebit Have you tested this on macOS Mojave? When I run the app using `electron electron/main.js` it works fine. But when I build the app using `electron-builder --win --mac` it stops working. The app always restores on the main monitor.

I have also tested this with the electron-boilerplate app, and the issue doesn't occur. It works perfectly.

MathieuDebit commented:
Sure I tested it on OSX Mojave and Windows 10, I also published a version on my fork (https://www.npmjs.com/package/@getstation/electron-window-state). It builds correctly.

Do you have enough material to open a new issue?

tobiasmuecksch commented:
I don't have problems with building. My problem is, that the app won't restore the position correctly, if I install and run the built version.

MathieuDebit commented:
Hmm that's weird 🤔I don't know how this could happen, but we can handle this in another issue with reproduction steps and I'll look into it!

tobiasmuecksch commented:
Thanks. I'll create a demo repo and open an issue
